infrastructure/src/main/java/org/corfudb/infrastructure/CompactorLeaderServices.java

```diff
@@ -0,0 +1,497 @@
+        this.corfuStore = new CorfuStore(corfuRuntime);
+        this.nodeEndpoint = nodeEndpoint;
+        syslog = LoggerFactory.getLogger("syslog");
+
+        try {
+            this.compactionManagerTable = this.corfuStore.openTable(CORFU_SYSTEM_NAMESPACE,
+                    DistributedCompactor.COMPACTION_MANAGER_TABLE_NAME,
+                    StringKey.class,
+                    CheckpointingStatus.class,
+                    null,
+                    TableOptions.fromProtoSchema(CheckpointingStatus.class));
+
+            this.checkpointingStatusTable = this.corfuStore.openTable(CORFU_SYSTEM_NAMESPACE,
+                    DistributedCompactor.CHECKPOINT_STATUS_TABLE_NAME,
+                    TableName.class,
+                    CheckpointingStatus.class,
+                    null,
+                    TableOptions.fromProtoSchema(CheckpointingStatus.class));
+
+            this.activeCheckpointsTable = this.corfuStore.openTable(CORFU_SYSTEM_NAMESPACE,
+                    DistributedCompactor.ACTIVE_CHECKPOINTS_TABLE_NAME,
+                    TableName.class,
+                    ActiveCPStreamMsg.class,
+                    null,
+                    TableOptions.fromProtoSchema(ActiveCPStreamMsg.class));
+
+            this.checkpointTable = this.corfuStore.openTable(CORFU_SYSTEM_NAMESPACE,
+                    DistributedCompactor.CHECKPOINT,
+                    StringKey.class,
+                    RpcCommon.TokenMsg.class,
+                    null,
+                    TableOptions.fromProtoSchema(RpcCommon.TokenMsg.class));
+
+        } catch (Exception e) {
+            syslog.error("Caught an exception while opening Compaction management tables ", e);
+        }
+    }
+
+    /**
+     * The leader initiates the Distributed checkpointer
+     * Trims the log till 'trimtoken' saved in the checkpoint table
+     * Mark the start of the compaction cycle and populate CheckpointStatusTable with all the
+     * tables in the registry.
+     *
+     * @return
+     */
+    @VisibleForTesting
+    public boolean trimAndTriggerDistributedCheckpointing() {
+        syslog.info("=============Initiating Distributed Checkpointing============");
+        if (!isLeader) {
+            return false;
+        }
+        trimLog();
+        if (DistributedCompactor.isCheckpointFrozen(corfuStore, checkpointTable)) {
+            syslog.warn("Will not start checkpointing since checkpointing has been frozen");
+            return false;
+        }
+
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            CorfuStoreEntry compactionManagerRecord = txn.getRecord(
+                    DistributedCompactor.COMPACTION_MANAGER_TABLE_NAME, DistributedCompactor.COMPACTION_MANAGER_KEY);
+            CheckpointingStatus managerStatus = (CheckpointingStatus) compactionManagerRecord.getPayload();
+
+            if (managerStatus != null && (managerStatus.getStatus() == StatusType.STARTED ||
```

Comment:
getPayload MUST always return an enum, please change the code accordingly.

Reply:
CheckpointingStatus is not an enum in the first place. 
Also getPayload cannot return a default value as it needs to be changed here - https://github.com/CorfuDB/CorfuDB/blob/master/runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java#L453, and that will affect a lot of other callers.